### PR TITLE
[Del] Remove DistributedTensor 

### PR DESCRIPTION
**Pull request by yhna940**
## Title

- [Del] Remove DistributedTensor 

## Description

- This PR involves the removal of the `DistributedTensor` component from our codebase. This decision was made as the element, which overrode the `__torch_function__`, was identified to be the source of bugs across multiple operators. All sections of the code associated with `DistributedTensor` have been meticulously purged to ensure the eradication of these bugs. 

## Linked Issues

- https://github.com/hpcaitech/ColossalAI/issues/3752

### CODE REVIEW

File: oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -111,15 +100,14 @@ def __init__(
             hidden_dim=hidden_dim,
             search_range_mb=search_range_mb,
             min_chunk_size_mb=min_chunk_size_mb,
-            strict_ddp_flag=strict_ddp_mode,
         )
         self.heterogeneous_manager = HeterogeneousMemoryManager(
             placement_policy, self.chunk_manager, memstats
         )
         self.force_outputs_fp32 = force_outputs_fp32
         self.param_op_hook = HeterogeneousZeROHook(self.heterogeneous_manager)
```
Comment (KKIEEK) [inline oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py:108]: As far as I remember, `HeterogeneousZeROHook` was derived from `DistributedParamOpHook`, but it seems like you removed `param_op_hook.py`. Did it work?

File: oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -232,31 +220,29 @@ def _post_backward(self):
         self.heterogeneous_manager.post_iter()
 
     def grad_handle(self, p, grad):
+        self.param_op_hook.post_backward([p])
```
Comment (KKIEEK) [inline oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py:223]: What do you think about moving this line to `_post_backward` method.

File: oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py (hunk 1 above)
Comment (yhna940) [reply oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py:108]: Yes, that's correct. I have made a change to directly execute the hooks that were previously driven by the param hook manager. (In the past, this process was triggered by `__torch_function__` of the dist tensor.)

After the modification, it has been confirmed that there are no issues with the functioning.

File: oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py (hunk 2 above)
Comment (yhna940) [reply oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py:223]: Based on the flow of method invocations, it seems that the current implementation is correct:

_module's _pre_backward -> ZeRO hook's pre_backward -> backward -> ZeRO hook's post_backward -> parameter hook (grad_handle) -> _module's _post_backward

Before grad_handle is triggered, the ZeRO hook's post_backward should appropriately mark the tensor states.

Is that correct? If you identify any errors, please let me know, and I will make the necessary adjustments.

File: oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py (hunk 1 above)
Comment (KKIEEK) [reply oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py:108]: I mean, it doesn't look like the parent class of `HeterogeneousZeROHook` is defined.

Comment (yhna940) [reply oslo/torch/nn/parallel/data_parallel/zero/fully_sharded_data_parallel.py:108]: Thanks, I'll fix it right away. sorry.

**Pull request merged**